vermouth/ffinput.py

```diff
@@ -51,196 +55,275 @@
+        wildcard_section = tuple(self.section[:-1] + ['*'])
+        end_section = []
+        if self.section:
+            end_section = self.section[-1]
+        if tuple(self.section) in self.METH_DICT:
+            method, kwargs = self.METH_DICT[tuple(self.section)]
+        elif wildcard_section in self.METH_DICT and end_section not in self.METH_DICT:
+            method, kwargs = self.METH_DICT[wildcard_section]
+        else:
+            raise IOError("Can't parse line {} in section '{}' because the "
+                          "section is unknown".format(lineno, self.section))
+        try:
+            return method(self, line, lineno, **kwargs)
+        except Exception as error:
+            raise IOError("Problems parsing line {}. I think it should be a "
+                          "'{}' line, but I can't parse it as such."
+                          "".format(lineno, self.section)) from error
+
+    def finalize(self, lineno=0):
+        for block in self.blocks.values():
+            # Because of hos they are described in gromacs, proper and improper
+            # dihedral angles are all under the [ dihedrals ] section. However
+            # the way they are treated differently in the library, at least on how
+            # they generate edges. Here we move the all the impropers into their own
+            # [ impropers ] section.
+            propers = []
+            impropers = []
+            for dihedral in block.interactions.get('dihedrals', []):
+                if dihedral.parameters and dihedral.parameters[0] == '2':
+                    impropers.append(dihedral)
+                else:
+                    propers.append(dihedral)
+            block.interactions['dihedrals'] = propers
+            block.interactions['impropers'] = impropers
+            block.make_edges_from_interactions()
+        for link in self.links:
+            link.make_edges_from_interactions()
+        self.force_field.blocks.update(self.blocks)
+        self.force_field.links.extend(self.links)
+        modifications = {mod.name: mod for mod in self.modifications}
+        self.force_field.modifications.update(modifications)
+        super().finalize(lineno=lineno)
+
+    def get_context(self, context_type):
+        possible_contexts = {
+            'block': self.current_block,
+            'link': self.current_link,
+            'molmeta': self.current_link,
+            'modification': self.current_modification,
+        }
+        return possible_contexts[context_type]
+
+    def has_context(self):
+        open_contexts = [
+            self.current_block, self.current_link, self.current_modification]
+        return open_contexts != ([None] * len(open_contexts))
+
+    def _new_block(self):
+        self.current_block = Block(force_field=self.force_field)
+
+    def _new_link(self):
+        self.current_link = Link(force_field=self.force_field)
+        self.links.append(self.current_link)
```

Comment:
Why not finish the existing block when making a new one? Why/how are links and blocks different?

Reply:
Yes this is right. In fact the blocks get finished in the '[moleculetype]' line parser and the links here, all while the finalize_section method is completely empty. I suggest to move both to the finalize section method.

Reply:
Links do not need a name to be added to the list. Blocks are added to the dict as soon as the name is known.

Reply:
But is that necessary though? You can keep the self.current_block until the section is finished can't you? @pckroon  and I just spent some time tracking down to see how the information is actually added to the block dict. and the current setup makes it very obscure especially because links and blocks are treated differently and the finalize section method is completely empty.

Reply:
done